pkg/porter/explain.go

```diff
@@ -0,0 +1,291 @@
+package porter
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/deislabs/cnab-go/bundle"
+	"github.com/deislabs/porter/pkg/context"
+	"github.com/deislabs/porter/pkg/printer"
+	"github.com/olekukonko/tablewriter"
+	"github.com/pkg/errors"
+)
+
+type ExplainOpts struct {
+	BundleLifecycleOpts
+	printer.PrintOptions
+}
+
+// PrintableBundle holds a subset of pertinent values to be explained from a bundle.Bundle
+type PrintableBundle struct {
+	Name        string                         `json:"name" yaml:"name"`
+	Description string                         `json:"description,omitempty" yaml:"description,omitempty"`
+	Version     string                         `json:"version" yaml:"version"`
+	Parameters  map[string]PrintableParameter  `json:"parameters,omitempty" yaml:"parameters,omitempty"`
+	Credentials map[string]PrintableCredential `json:"credentials,omitempty" yaml:"credentials,omitempty"`
+	Outputs     map[string]PrintableOutput     `json:"outputs,omitempty" yaml:"outputs,omitempty"`
+	Actions     map[string]PrintableAction     `json:"customActions,omitempty" yaml:"customActions,omitempty"`
+}
+
+type PrintableCredential struct {
+	Description string `json:"description" yaml:"description"`
+	Required    bool   `json:"required" yaml:"required"`
+}
+
+type PrintableOutput struct {
+	Type        interface{} `json:"type" yaml:"type"`
+	ApplyTo     string      `json:"applyTo" yaml:"applyTo"`
+	Description string      `json:"description" yaml:"description"`
+}
+
+type PrintableParameter struct {
+	Type        interface{} `json:"type" yaml:"type"`
+	Default     interface{} `json:"default" yaml:"default"`
+	ApplyTo     string      `json:"applyTo" yaml:"applyTo"`
+	Description string      `json:"description" yaml:"description"`
+	Required    bool        `json:"required" yaml:"required"`
+}
+
+type PrintableAction struct {
+	Modifies bool `json:"modifies" yaml:"modifies"`
+	// Stateless indicates that the action is purely informational, that credentials are not required, and that the runtime should not keep track of its invocation
+	Stateless bool `json:"stateless" yaml:"stateless"`
+	// Description describes the action as a user-readable string
+	Description string `json:"description" yaml:"description"`
+}
+
+func (o *ExplainOpts) Validate(args []string, cxt *context.Context) error {
+	err := o.sharedOptions.Validate(args, cxt)
+	if err != nil {
+		return err
+	}
+	err = o.ParseFormat()
+	if err != nil {
+		return err
+	}
+	if o.Tag != "" {
+		o.File = ""
+		o.CNABFile = ""
+
+		return o.validateTag()
+	}
+	return nil
+}
+
+func (p *Porter) Explain(o ExplainOpts) error {
+	err := p.prepullBundleByTag(&o.BundleLifecycleOpts)
+	if err != nil {
+		return errors.Wrap(err, "unable to pull bundle before invoking credentials generate")
+	}
+
+	err = p.applyDefaultOptions(&o.sharedOptions)
+	if err != nil {
+		return err
+	}
+	err = p.ensureLocalBundleIsUpToDate(o.bundleFileOptions)
```

Comment:
I'm thinking this call is not necessary, right?  (Would re-build if bundle differed/out-of-date with local manifest.)

Reply:
Actually, even though perhaps a main use case would be to explain a remote bundle, I can also see a use case for explaining a local bundle... in that scenario, we might as well would want to ensure it'd be up-to-date.  Therefore, seems we'd want to keep this!

Reply:
Yeah, we for sure want to be able to use this on canonical JSON bundle files that are supper hard to read or a porter.yaml that people might have shared with others.